Codegen: streamline code for subproject generation

thockin opened the pull request:
1) change "all" into explicit lists of generators
2) make generate-internal-groups be a superset of generate-groups
4) call internal from generate-groups

I also have an idea to followup with a new generate-kube-code.sh script which is more friendly and automatic (maybe no more open-coded lists of versions), and convert callers to that (and deprecate old scripts)

There will be several followups to this.  One that I'd like to pursue is getting rid of the update-codegen bash code which deals with indirect arrays for openapi - specifically for subprojects.  Several of them *want* openapi generation, but the generate-groups logic does not produce the same results.  I don't know enough about it to dig in yet.

/kind cleanup

```release-note
NONE
```

liggitt commented on staging/src/k8s.io/code-generator/generate-groups.sh:
```diff
@@ -18,14 +18,15 @@ set -o errexit
 set -o pipefail
 
-# generate-groups generates everything for a project with external types only, e.g. a project based
+# generate-groups generates everything for a project with external types, e.g. a project based
```
there are *lots* of people calling this for [APIs which have no internal types defined](https://cs.github.com/?scopeName=All+repos&scope=&q=%2F%5C%2Fgenerate-groups.sh+%22%3Fall%2F). I'm pretty sure the conversion generator won't work for them, and I'm unsure about the other bits now included in `all`

thockin replied:
I question whether "all" is useful any more.  I could force it to be the historical meaning and just freeze it.  Or we could make it an error and tell people "replace 'all' with ..." locking in the historical meaning.

Prefs?

liggitt replied:
The only opinions I have:
* if existing invocations are going to stop working, they should stop working in as clear and obvious a way as possible
* we should be able to recommend *something* to people who previously generated conversions for API types in two locations using this script

Other than that, I don't feel strongly about the shape this evolves into. @sttts might feel differently.

liggitt commented on staging/src/k8s.io/code-generator/generate-internal-groups.sh:
```diff
@@ -61,40 +61,34 @@ gobin="${GOBIN:-$(go env GOPATH)/bin}"
 for GVs in ${GROUPS_WITH_VERSIONS}; do
   IFS=: read -r G Vs <<<"${GVs}"
 
-  if [ -n "${INT_APIS_PKG}" ]; then
-    ALL_FQ_APIS+=("${INT_APIS_PKG}/${G}")
-  fi
```
also, dropping adding the `$INTERNAL_API_PKG/$GROUP` directory to $ALL_FQ_APIS means that deepcopy-gen and conversion-gen aren't getting invoked with knowledge of the internal type packages... how are those generators working properly?

thockin replied:
Deepcopy doesn't need the peer API.  Conversion does but each input (versioned) API explicitly calls it out (e.g. `+k8s:conversion-gen=k8s.io/apiextensions-apiserver/pkg/apis/apiextensions` in doc.go)

Effectively there is no such thing as "internal" any more.  Internal clients don't (seem to?) make sense.  Conversion is hgandled.  I'll have to look at those other repos - I was really only looking at internal users.  Good idea.

thockin reviewed (commented):
Will push with the args-numbering fix.

Will test external repos.

thockin commented:
Looking at a few of these external projects - they are all over the place.  The expansion of "all" certainly impacts many of them, but also many of them show deltas against HEAD *before* this PR.

Some of them *do* generate against internal APIs, but not clients - deepcopy is the one that popped up.

So the question I have is this:

I find the overlap between the generate-groups and generate-internal groups to be hard to understand.  I could just hold my nose and leave it alone (maybe declare both to be deprecated and encourage people to DIY use the binaries from the code-generator repo) .  Or I could try to reduce "internal" to just be things that make sense for internal (e.g. just deepcopy).  

I'll chew on it, thoughts are welcome

fedebongio commented:
/triage accepted

thockin committed:
Trivial change to reduce diffs

This just makes it easier to diff generate-groups and
generate-internal-groups

thockin committed:
Codegen subprojects: reify 'all' into explicit

thockin committed:
Codegen subprojects: regen defaulters when needed

If you manually remove generated defaults, they do not get regenerated
without this.

thockin committed:
Codegen metrics: include v1beta2

This is why these scripts are terrible - open-coded lists always drift.
We will have to tackle that later.

thockin committed:
Codegen subprojects: deprecate "all" in scripts

Replace "all" with the historical meaning of it in the generate-groups
scripts.

thockin committed:
Codegen subprojects: consolidate generate-*groups

They both still exist, with the same API as before, but now
generate-groups calls generate-internal-groups, which is a superset.

thockin commented:
OK, I took this in a different direction.

1) change "all" into explicit lists of generators
2) make generate-internal-groups be a superset of generate-groups
3) call internal from generate-groups

If this flies better, I will rebase other PRs onto it.

I also have an idea to followup with a new `generate-kube-code.sh` script which is more friendly and automatic (maybe no more open-coded lists of versions), and convert callers to that (and deprecate old scripts)

thockin commented:
/retest

liggitt commented on staging/src/k8s.io/code-generator/generate-groups.sh:
```diff
@@ -25,7 +25,7 @@ if [ "$#" -lt 4 ] || [ "${1}" == "--help" ]; then
   cat <<EOF
 Usage: $(basename "$0") <generators> <output-package> <apis-package> <groups-versions> ...
 
-  <generators>        the generators comma separated to run (deepcopy,defaulter,applyconfiguration,client,lister,informer) or "all".
```
it's so squirrelly defaulter-gen is in generate-internal-groups... logically, defaulting belongs in the external group generator, and doesn't need the internal types *at all* (defaulting only looks at the external type, and is per-version)

thockin replied:
I could flip it around so internal call non-internal.  The point is that there's no real reason to have both and these scripts should EOL:

I have more commits queued up, some choice ones:
https://github.com/kubernetes/kubernetes/commit/692adf02d8c35554e0a8838ea32c93837f50796e
https://github.com/kubernetes/kubernetes/commit/3a013a771fd0a5dc2d37a9495acf4cbb3e8de5ed

See the whole branch if you want, I was just rate-limiting how many PRs I send you and end up needing to rebase :)

https://github.com/kubernetes/kubernetes/compare/master...thockin:kubernetes:codegen_new_generate_groups_approach

liggitt replied:
Ah, just got to the end of the stack, if this is a somewhat intermediate state and it makes more sense in the end, I don't especially care, was just noting things that looked weird.

it's not like this is *removing* defaulter-gen from the external script, so I don't object to making what this *does* do explicit

liggitt commented on staging/src/k8s.io/code-generator/generate-groups.sh:
```diff
@@ -46,6 +49,15 @@ APIS_PKG="$3"
+    ALL="applyconfiguration,client,deepcopy,informer,lister"
+    echo "WARNING: Specifying \"all\" as a generator is deprecated."
+    echo "WARNING: Please list the specific generators needed."
+    echo "WARNING: \"all\" is now an alias for \"${ALL}\""
```
maybe call out that it will not include new generators added to this list, like `defaulter`?

thockin replied:
ACK, will push

liggitt commented on staging/src/k8s.io/code-generator/generate-internal-groups.sh:
```diff
@@ -64,7 +64,18 @@ fi
+      openapi-gen
+  )
+  # shellcheck disable=2046 # printf word-splitting is intentional
+  GO111MODULE=on go install $(printf "k8s.io/code-generator/cmd/%s " "${BINS[@]}")
```
is a single `go install` call more efficient in a way that matters (I don't know if it shares resolution of deps, or if that is cheap enough to not matter)

thockin replied:
mostly just simpler

liggitt commented on staging/src/k8s.io/code-generator/generate-internal-groups.sh:
```diff
@@ -64,7 +64,18 @@ fi
+  GO111MODULE=on go install $(printf "k8s.io/code-generator/cmd/%s " "${BINS[@]}")
 )
 
 # Go installs the above commands to get installed in $GOBIN if defined, and $GOPATH/bin otherwise:
```
I didn't diff the external/internal invocations... all the options were identical?

thockin replied:
yes

liggitt commented:
/lgtm
/hold

a couple suggestions about more explicit warning messages and a question about `go install`, but looks reasonable as-is, feel free to unhold / defer the suggestions

thockin committed:
Codegen: new tools will not become part of 'all'

liggitt commented:
/lgtm
